atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresDdlTable.java

```diff
@@ -118,10 +126,67 @@ public void checkDatabaseVersion() {
 
     @Override
     public void compactInternally() {
+        if (compactionSemaphore.tryAcquire()) {
+            try {
+                checkLastCompactionTimeAndRunCompact();
+            } finally {
+                compactionSemaphore.release();
+            }
+        }
+    }
+
+    private void checkLastCompactionTimeAndRunCompact() {
+        if (config.compactInterval().toMilliseconds() == 0 || (config.compactInterval().toMilliseconds() > 0
```

Comment:
1. This conditional is pretty ugly. It would probably be useful instead to just write a function that returns a boolean of whether or not compaction should run.

2. The way you've written it you do two round trips to the DB just to figure out the time since the last compaction. This can be done pretty trivially as a single round trip.

3. There is little reason to construct/use Timestamp classes -- longs (defined as utc milliseconds) work fine, especially since this is completely done internally in this class.  The way you get the DB to give you a numeric value is to use `SELECT EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000` (note this can give you a fractional value so use `FLOOR` to get an integer value for this).  You can also use the SQL function `GREATEST` to get the largest value out of all of the columns.

4. Are we sure that the pg_stat_user_tables gets updated at the start of a vacuum run? If it doesn't update until afterwards, this table doesn't help you, and you might need to make your own table to keep track of when vacuums get started.

```
@Override
public void compactInternally() {
   if (compactionSemaphore.tryAcquire()) {
       try {
           if (shouldRunCompaction()) {
               runCompactOnTable();
           }
       } finally {
           compactionSemaphore.release();
       }
   }
}

private boolean shouldRunCompaction() {
    long compactIntervalMillis = config.compactInterval().toMilliseconds();
    if (compactIntervalMillis <= 0) {
        return true;
    }

    long millisSinceLastCompact = getMillisSinceLastCompact();
    return millisSinceLastCompact >= compactIntervalMillis;
}

private long getMillisSinceLastCompact() {
    AgnosticResultSet rs = conns.get().selectResultSetUnregisteredQuery(
          "SELECT FLOOR(EXTRACT(EPOCH FROM GREATEST( " +
              + "  last_vacuum, last_autovacuum, last_analyze, last_autoanalyze"
              + "))*1000) AS last, "
              + "FLOOR(EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000) AS current "
              + "FROM pg_stat_user_tables WHERE relname = ?",
        prefixedTableName());

    AgnosticResultRow row = Iterables.getOnlyElement(rs.rows(), null);
    if (row == null) {
        // query could have no results
        // effectively the last vacuum was infinite time ago, so use Long.MAX_VALUE as the time since then
        return Long.MAX_VALUE;
    }

    return row.getLong("current") - row.getLong("last");
}
```

Reply:
Thanks @berler for the review!

Agree with 1, 2 and 3. Will make the changes after adding good tests. Due to lack of familiarity with Postgres queries, I had inefficient and very simple queries in there, thanks for the suggestions.

For 4:
I think we update the timestamps approximately at the start (upto after 500ms) of a Vacuum. Background sweep should generally take more that that time to return to compacting a table.

The only case where the approximate stat update might result in multiple vacuums is if we are kicking off sweeps via the endpoint (but that's not very common). We are trying to solve the issue where there could be 10-15 vacuums queued up on the same table, and this should prevent it in most cases.

The stats are updated based on the Postgres stats collector frequency (500ms by default, but configurable) and we will have stats for the current running processes if `track_activities` is true (true by default, but configurable). Do you think these parameters might be configured differently in the general case?

Reply:
I was trying to test the query above and found this:

```
atlas=# SELECT relname, last_vacuum, last_autovacuum, last_analyze, last_autoanalyze from pg_stat_user_tables where relname = 'ns__pt_kvs_test';
 ns__pt_kvs_test |             |                 |              | 2017-12-10 02:51:31.190146+00

atlas=# SELECT CURRENT_TIMESTAMP;                                                                                                                                                                      2017-12-10 03:04:57.472697+00

atlas=#  SELECT FLOOR(EXTRACT(EPOCH FROM GREATEST(last_vacuum, last_autovacuum, last_analyze, last_autoanalyze))*1000) AS last, FLOOR(EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000) AS current FROM pg_stat_user_tables WHERE relname = 'ns__pt_kvs_test';
 1512874291190 | 1512875103905

```

When no Vacuum has been run we still have a value for the last timestamp `1512874291190`. Also, we loose testing for different values of the four kinds of Vacuum timestamps if we have the consolidated logic making it difficult to catch bugs in unit tests. I dont feel having the greatest timestamp calculation logic in the DB saves us much except for traffic on the wire, this call is very in-frequent i.e. at the end of sweeping one entire table, so wouldn't invest in optimizing for this.
I can change this to have one DB Call instead of two and then parse the timestamp objects.

Reply:
Ignore the previous comment, the `resultSet` returned has a `null` value if no vacuum has been performed, using the above specified query for this. We do loose some granularity for the testing but this is the trade-off we have to make for optimized queries.